### PR TITLE
feat(sync): add StateSyncClient and SyncBlock

### PR DESCRIPTION
**Pull request by ShahakShama**

### CODE REVIEW

**Review by ShahakShama — COMMENTED** [review]
+reviewer:@alonh5 

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007#-:-OBZqvYFCT-s5-sFwzzE:b-wqqsco)* status: 0 of 7 files reviewed, all discussions resolved (waiting on @alonh5, @dan-starkware, @eitanm-starkware, and @Itay-Tsabary-Starkware)

**Review by Itay-Tsabary-Starkware — APPROVED** [review]
:lgtm:

> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007#-:-OBZtfyp777P6FkhH9b9:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @alonh5, @dan-starkware, and @eitanm-starkware)

**Review by Itay-Tsabary-Starkware — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @alonh5, @dan-starkware, and @eitanm-starkware)

___
*[`crates/starknet_state_sync_types/src/communication.rs` line 16 at r1](https://reviewable.io/reviews/starkware-libs/sequencer/2007#-OBZtrsNDTHLp-VmGz3L:-OBZtrsNDTHLp-VmGz3M:b-guvphp) ([raw file](https://github.com/starkware-libs/sequencer/blob/a1b3475fb0820c818eb692c987d08ba43d4d266a/crates/starknet_state_sync_types/src/communication.rs#L16)):*
> ```XML
>     ) -> StateSyncClientResult<Option<SyncBlock>>;
> 
>     // TODO: Add state reader methods for gateway.
> ```

Please use named TODOs, throughout (non-blocking)



_Code quote:_
```XML
// TODO: Add state reader methods for gateway.
```

**Review by alonh5 — APPROVED** [review]
:lgtm:

> Reviewed 3 of 7 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007#-:-OB_av_KE5DWuAflfWy6:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @dan-starkware, @eitanm-starkware, and @ShahakShama)

___
*[`crates/starknet_state_sync_types/src/state_sync_types.rs` line 13 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/2007#-OB__Dt11GEB4uKlhXnK:-OB__Dt11GEB4uKlhXnL:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/f7c947265a12d268d406b25b54d669ec7b56ae4c/crates/starknet_state_sync_types/src/state_sync_types.rs#L13)):*
> ```XML
> pub type StateSyncResult<T> = Result<T, StateSyncError>;
> 
> /// A block that came from state sync.
> ```





_Suggestion:_
```XML
/// A block that came from a/the state sync.
```
___
*[`crates/starknet_state_sync_types/src/state_sync_types.rs` line 16 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/2007#-OB_a3448s3CgdeTPvJg:-OB_a3456d1yg8dbHVvG:undefined) ([raw file](https://github.com/starkware-libs/sequencer/blob/f7c947265a12d268d406b25b54d669ec7b56ae4c/crates/starknet_state_sync_types/src/state_sync_types.rs#L16)):*
> ```XML
> /// Contains all the data needed to update the state of the system about this block.
> ///
> /// We trust blocks that came from sync. Therefore, SyncBlock doesn't contain data needed for
> ```





_Suggestion:_
```XML
Blocks that came from the state sync are trusted
```
___
*[`crates/starknet_state_sync_types/src/state_sync_types.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/2007#-OB_aTCXDzDwP4uPw79N:-OB_aTCXDzDwP4uPw79O:bat4myo) ([raw file](https://github.com/starkware-libs/sequencer/blob/f7c947265a12d268d406b25b54d669ec7b56ae4c/crates/starknet_state_sync_types/src/state_sync_types.rs#L26)):*
> ```XML
>     pub deprecated_classes: HashMap<ClassHash, DeprecatedContractClass>,
>     pub transaction_hashes: Vec<TransactionHash>,
> }
> ```

The batcher only needs the state_diff and the transaction_hashes. Are all of the other fields needed by the consensus? @dan-starkware 



_Code quote:_
```XML
pub struct SyncBlock {
    pub block_hash: BlockHash,
    pub parent_block_hash: BlockHash,
    pub state_diff: ThinStateDiff,
    pub classes: HashMap<ClassHash, ContractClass>,
    // TODO: Consider removing deprecated classes.
    pub deprecated_classes: HashMap<ClassHash, DeprecatedContractClass>,
    pub transaction_hashes: Vec<TransactionHash>,
}
```

File: crates/starknet_state_sync_types/src/state_sync_types.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+use crate::errors::StateSyncError;
+
+pub type StateSyncResult<T> = Result<T, StateSyncError>;
+
+/// A block that came from state sync.
+/// Contains all the data needed to update the state of the system about this block.
+///
+/// We trust blocks that came from sync. Therefore, SyncBlock doesn't contain data needed for
+/// verifying the block
+pub struct SyncBlock {
```
[the remark this answers is not in the document]
Comment (ShahakShama) [reply crates/starknet_state_sync_types/src/state_sync_types.rs:16]: :+1:

**Review by ShahakShama — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007)* status: 6 of 7 files reviewed, 2 unresolved discussions (waiting on @alonh5, @AlonLStarkWare, @dan-starkware, @eitanm-starkware, and @graphite-app[bot])

**Review by alonh5 — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007)* status: all files reviewed, 2 unresolved discussions (waiting on @AlonLStarkWare, @dan-starkware, @eitanm-starkware, @graphite-app[bot], and @ShahakShama)

**Comment by dan-starkware** [comment]
*[`crates/starknet_state_sync_types/src/state_sync_types.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/2007#-OB_aTCXDzDwP4uPw79N:-OBeHnPr3L4ySPKpLQ_K:b-c53he7) ([raw file](https://github.com/starkware-libs/sequencer/blob/f7c947265a12d268d406b25b54d669ec7b56ae4c/crates/starknet_state_sync_types/src/state_sync_types.rs#L26)):*
<details><summary><i>Previously, alonh5 (Alon Haramati) wrote…</i></summary><blockquote>

The batcher only needs the state_diff and the transaction_hashes. Are all of the other fields needed by the consensus? @dan-starkware
</blockquote></details>

We will probably need a subset of these later. IMO we should start with what is currently needed for batcher and add more when we need.
In general I prefer thin e2e approach and then adding more functionality

**Review by alonh5 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007)* status: all files reviewed, 2 unresolved discussions (waiting on @AlonLStarkWare, @dan-starkware, @eitanm-starkware, and @graphite-app[bot])

___
*[`crates/starknet_state_sync_types/src/state_sync_types.rs` line 26 at r2](https://reviewable.io/reviews/starkware-libs/sequencer/2007#-OB_aTCXDzDwP4uPw79N:-OBeKIOT5IHfmvNfnN2q:b-f4o2wu) ([raw file](https://github.com/starkware-libs/sequencer/blob/f7c947265a12d268d406b25b54d669ec7b56ae4c/crates/starknet_state_sync_types/src/state_sync_types.rs#L26)):*
<details><summary><i>Previously, dan-starkware wrote…</i></summary><blockquote>

We will probably need a subset of these later. IMO we should start with what is currently needed for batcher and add more when we need.
In general I prefer thin e2e approach and then adding more functionality
</blockquote></details>

Great, I agree. Shahak, can you remove the other fields for now?

**Review by alonh5 — APPROVED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007)* status: 6 of 7 files reviewed, 1 unresolved discussion (waiting on @AlonLStarkWare, @dan-starkware, @eitanm-starkware, and @graphite-app[bot])

**Review by alonh5 — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007)* status: all files reviewed, 1 unresolved discussion (waiting on @AlonLStarkWare, @dan-starkware, @eitanm-starkware, and @graphite-app[bot])

**Review by dorimedini-starkware — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007)* status: all files reviewed, 1 unresolved discussion (waiting on @AlonLStarkWare, @dan-starkware, @eitanm-starkware, and @graphite-app[bot])

**Review by meship-starkware — COMMENTED** [review]
> Dismissed @graphite-app[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @AlonLStarkWare, @dan-starkware, and @eitanm-starkware)

**Review by alonh5 — APPROVED** [review]
> Dismissed @graphite-app[bot] from a discussion.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/sequencer/2007)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @AlonLStarkWare, @dan-starkware, and @eitanm-starkware)

**Pull request merged**